From patchwork Thu Aug 30 00:54:54 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Peter Shier <pshier@google.com>
X-Patchwork-Id: 10581101
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 7DD095A4
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 00:55:10 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 6E3D42BA99
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 00:55:10 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 5FE592BAF3; Thu, 30 Aug 2018 00:55:10 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-15.5 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI,
	USER_IN_DEF_DKIM_WL autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 025A12BA99
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 30 Aug 2018 00:55:10 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1725904AbeH3Eyi (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 30 Aug 2018 00:54:38 -0400
Received: from mail-pg1-f202.google.com ([209.85.215.202]:43293 "EHLO
        mail-pg1-f202.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1725836AbeH3Eyi (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 30 Aug 2018 00:54:38 -0400
Received: by mail-pg1-f202.google.com with SMTP id u6-v6so4091044pgn.10
        for <kvm@vger.kernel.org>; Wed, 29 Aug 2018 17:55:07 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=google.com; s=20161025;
        h=date:message-id:mime-version:subject:from:to:cc;
        bh=nNsv/luKoJZLFHEq1xl34F6tPuz10wS7StuzkqSovi4=;
        b=lu3PJunmiJMRfga0GJ5z/6rEPdo3Xt9RXfxO567gMa+7aJa1PPTojcRgAuwFlj+OGK
         LSgrCRJKyHZGbUUe6LeLbKjk4telNeVd9869KAmm4lSmshuKaTRDXGNs2AI728XP+N5I
         3h0B2V9P3TbWdFLy1G3yDZQ/MpvNy3xnYg6gTm2imlsR4loA6dyCrMMH4b/Q6IN4BSEj
         MkVvJXl/stpRWd4Ly6w6S7CpqUMbhCBQdIWbD4qTEZdbJ390sR3AgaNffLB16oN3nS4p
         jpLok8lIUTZbN7ec3k/2x3aVeE4WRs6IKUsCu8KBnDeaNab/z2uHHj3ddpw3yN2es+Zu
         y4LA==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:message-id:mime-version:subject:from:to:cc;
        bh=nNsv/luKoJZLFHEq1xl34F6tPuz10wS7StuzkqSovi4=;
        b=g9FEg4779pKuZnfT0TL6MUGlR7dupFjpY0cRfVlJNH7/S41VEPwmjqv1flL7u7OCmB
         ROWl632Ugi1KS1hS2vBNpoIG4mHw67LUsS+tHNnwSsLoyudWt8q8HhEZVa1/Rf3sbRzd
         pxaanQbCaVyK6bPHJKDeB3gOcKIBG5vGZEuHZN+sP+RmtV+c4vdKYeEzj4y0stqtCubh
         NSA/dRuHvpJ68iItewjVc1s2GPOkakBzMIjh/VcLiUxht1/ZTcbeS5RsG4uQzt70PI+F
         7irsb/bo7MjtQdUBOXh5K9XmWP9B1VR13MRLI2iz1pO9iXl3YQpy1u6si04UAyZnhGSV
         LcRw==
X-Gm-Message-State: APzg51CryGqrP43LCixgL4EKWxrqJ1zXmkoUNCxCUUdkAgNfBZ+tHlSK
        AqlNwpejGXWYHMoMGXZiILJPYfZhU1zEF7sCx5m1DgCDOwDC1OSu7YjoKiF8znNdaLHTaKbgMmr
        YIy2kOey6qFsYpurYjkOqVFTAt8hY52TlymmwkII1N7p1B2ZihMQdTbg3TA==
X-Google-Smtp-Source: 
 ANB0VdZVVwD3HPGQEJ+614b8N2g3reF/sQNXb3SXeUfAtcwc2ETgSYaPHsAfU+WswSgraj8mt0+F8J42zE0=
X-Received: by 2002:a63:3dcd:: with SMTP id
 k196-v6mr2247046pga.37.1535590507334;
 Wed, 29 Aug 2018 17:55:07 -0700 (PDT)
Date: Wed, 29 Aug 2018 17:54:54 -0700
Message-Id: <20180830005454.142075-1-pshier@google.com>
Mime-Version: 1.0
X-Mailer: git-send-email 2.19.0.rc0.228.g281dcd1b4d0-goog
Subject: [PATCH v2] KVM: Remove CREATE_IRQCHIP/SET_PIT2 race
From: Peter Shier <pshier@google.com>
To: kvm@vger.kernel.org
Cc: Steve Rutherford <srutherford@google.com>,
        Peter Shier <pshier@google.com>,
        Jim Mattson <jmattson@google.com>
Content-Type: text/plain; charset="UTF-8"
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

From: Steve Rutherford <srutherford@google.com>

Fixes a NULL pointer dereference, caused by the PIT firing an interrupt
before the interrupt table has been initialized.

SET_PIT2 can race with the creation of the IRQchip. In particular,
if SET_PIT2 is called with a low PIT timer period (after the creation of
the IOAPIC, but before the instantiation of the irq routes), the PIT can
fire an interrupt at an uninitialized table.

Signed-off-by: Steve Rutherford <srutherford@google.com>
Signed-off-by: Peter Shier <pshier@google.com>
Signed-off-by: Jim Mattson <jmattson@google.com>
---
 arch/x86/kvm/x86.c | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)

diff --git a/arch/x86/kvm/x86.c b/arch/x86/kvm/x86.c
index 506bd2b4b8bb7..188bc3fc907f1 100644
--- a/arch/x86/kvm/x86.c
+++ b/arch/x86/kvm/x86.c
@@ -4518,10 +4518,11 @@ long kvm_arch_vm_ioctl(struct file *filp,
 		r = -EFAULT;
 		if (copy_from_user(&u.ps, argp, sizeof u.ps))
 			goto out;
+		mutex_lock(&kvm->lock);
 		r = -ENXIO;
-		if (!kvm->arch.vpit)
-			goto out;
-		r = kvm_vm_ioctl_set_pit(kvm, &u.ps);
+		if (kvm->arch.vpit)
+			r = kvm_vm_ioctl_set_pit(kvm, &u.ps);
+		mutex_unlock(&kvm->lock);
 		break;
 	}
 	case KVM_GET_PIT2: {
@@ -4541,10 +4542,11 @@ long kvm_arch_vm_ioctl(struct file *filp,
 		r = -EFAULT;
 		if (copy_from_user(&u.ps2, argp, sizeof(u.ps2)))
 			goto out;
+		mutex_lock(&kvm->lock);
 		r = -ENXIO;
-		if (!kvm->arch.vpit)
-			goto out;
-		r = kvm_vm_ioctl_set_pit2(kvm, &u.ps2);
+		if (kvm->arch.vpit)
+			r = kvm_vm_ioctl_set_pit2(kvm, &u.ps2);
+		mutex_unlock(&kvm->lock);
 		break;
 	}
 	case KVM_REINJECT_CONTROL: {
